### PR TITLE
Add `FilteredAccess::empty` and simplify the implementatin of `update_component_access` for `AnyOf`/`Or`

### PR DESCRIPTION
**Pull request by SkiFire13**
# Objective

- The implementation of `update_component_access` for `AnyOf`/`Or` is kinda weird due to special casing the first filter, let's simplify it;
- Fundamentally we want to fold/reduce the various filters using an OR operation, however in order to do a proper fold we need a neutral element for the initial accumulator, which for OR is FALSE. However we didn't have a way to create a `FilteredAccess` value corresponding to FALSE and thus the only option was reducing, which special cases the first element as being the initial accumulator.

This is an alternative to https://github.com/bevyengine/bevy/pull/14026

## Solution

- Introduce `FilteredAccess::empty` as a way to create a `FilteredAccess` corresponding to the logical proposition FALSE;
- Use it as the initial accumulator for the above operations, allowing to handle all the elements to fold in the same way.

---

## Migration Guide

- The behaviour of `AnyOf<()>` and `Or<()>` has been changed to match no archetypes rather than all archetypes to naturally match the corresponding logical operation. Consider replacing them with `()` instead.

### CODE REVIEW

File: crates/bevy_ecs/src/query/fetch.rs [hunk 1]
```diff
@@ -1862,24 +1862,18 @@ macro_rules! impl_anytuple_fetch {
...
-                        $name::update_component_access($name, &mut _new_access);
-                        _new_access.required = _access.required.clone();
-                        _not_first = true;
-                    }
+                    let mut intermediate = _access.clone();
+                    $name::update_component_access($name, &mut intermediate);
+                    _new_access.append_or(&intermediate);
                 )*
 
+                // we use an intermediate access because we only want to update the filter_sets, not the access
```
Comment (cBournhonesque) [inline crates/bevy_ecs/src/query/fetch.rs:1876]: I would put this above, when `intermediate` is created

File: crates/bevy_ecs/src/query/filter.rs [hunk 2]
```diff
@@ -443,21 +443,17 @@ macro_rules! impl_or_query_filter {
...
-                        _new_access.extend_access(&intermediate);
-                    } else {
-                        $filter::update_component_access($filter, &mut _new_access);
-                        _new_access.required = access.required.clone();
-                        _not_first = true;
-                    }
+                    let mut intermediate = access.clone();
+                    $filter::update_component_access($filter, &mut intermediate);
+                    _new_access.append_or(&intermediate);
+                    _new_access.extend_access(&intermediate);
```
Comment (cBournhonesque) [inline crates/bevy_ecs/src/query/filter.rs:452]: I think this line can be removed

File: crates/bevy_ecs/src/query/filter.rs [hunk 3]
```diff
@@ -443,21 +443,17 @@ macro_rules! impl_or_query_filter {
...
                 )*
 
+                _new_access.required = access.required.clone();
```
Comment (cBournhonesque) [inline crates/bevy_ecs/src/query/filter.rs:455]: I think this line can be removed as well; they were needed just because we had to copy the access to seed the accumulation

File: crates/bevy_ecs/src/query/fetch.rs (hunk 1 above)
Comment (SkiFire13) [reply crates/bevy_ecs/src/query/fetch.rs:1876]: It was up there initially, but while updating the method it didn't really made sense to me. `Or` also creates an `intermediate` but not because it only wants to update `filter_sets`, but because `access` should be kept intact for the next `update_component_access` call, and `_new_access` should be modified by calling `append_or`, not directly by the `update_component_access` of the current item.

Instead it's more interesting what happens right below this comment, as that's keeping only the `filter_sets` from the accumulated `_new_access`, instead of just assigning `_new_access` to `access`.

Comment (cBournhonesque) [reply crates/bevy_ecs/src/query/fetch.rs:1876]: Hm i see, you're commenting on why we need a `new_access`, because we don't want the call `update_component_access` to update the access. That makes sense to me.

I guess I wanted to have an additional comment for why we need a call to `let mut intermediate = _access.clone()` because it wasn't obvious to me at all at first. But I guess that's an independent concern, there's other parts of the codebase that use this 'trick' and it's not commented anywhere. Can leave it off for a future PR

**Review by cBournhonesque — APPROVED** [review]

File: crates/bevy_ecs/src/query/fetch.rs (hunk 1 above)
Comment (SkiFire13) [reply crates/bevy_ecs/src/query/fetch.rs]: I've updated the comments, let me know if they make more sense now.

File: crates/bevy_ecs/src/query/filter.rs (hunk 2 above)
Comment (SkiFire13) [reply crates/bevy_ecs/src/query/filter.rs:456]: I was also wondering about that, but then I noticed that filter can still access the change ticks of components, and this needs to conflict with other queries in the system that mutably access those change ticks.

File: crates/bevy_ecs/src/query/fetch.rs [hunk 4]
```diff
@@ -1861,30 +1861,29 @@ macro_rules! impl_anytuple_fetch {
...
-                        let mut intermediate = _access.clone();
-                        $name::update_component_access($name, &mut intermediate);
-                        _new_access.append_or(&intermediate);
-                    } else {
-                        $name::update_component_access($name, &mut _new_access);
-                        _new_access.required = _access.required.clone();
-                        _not_first = true;
-                    }
+                    // Create an intermediate because `access`'s value needs to be preserved
+                    // for the next query data, and `_new_access` has to be modified only by `append_or` to it.
```
Comment (cBournhonesque) [inline crates/bevy_ecs/src/query/fetch.rs:1872]:
```suggestion
                    // for the next query data, and `_new_access` has to be modified only by `append_or` to it, which only updates the `filter_sets`, not the `access`
```

File: crates/bevy_ecs/src/query/filter.rs (hunk 3 above)
Comment (SkiFire13) [reply crates/bevy_ecs/src/query/filter.rs]: The seed is still an empty `FilteredAccess`, just using that will lose informations about all the required components.

I can however avoid the `.clone()` here and instead use `std::mem::take`.

File: crates/bevy_ecs/src/query/filter.rs (hunk 2 above)
Comment (cBournhonesque) [reply crates/bevy_ecs/src/query/filter.rs:456]: I'm not sure I understand, but maybe you could add a test that fails if this line is missing?

Currently, if I remove this line, all tests are passing

Comment (SkiFire13) [reply crates/bevy_ecs/src/query/filter.rs:456]: I've added a test that fails when this line is commented.

TLDR: `Query<(), Or<(Changed<A>,)>>` has to conflict with `Query<&mut A>` because the first read the change trackers and the second one writes them. However if `_new_access.extend_access(&intermediate);` is omitted then the first one is considered as not reading anything and that makes them compatible.

**Review by chescock — APPROVED** [review]
This makes those implementations much clearer!

File: crates/bevy_ecs/src/query/access.rs [hunk 5]
```diff
@@ -353,6 +353,16 @@ impl<T: SparseSetIndex> From<FilteredAccess<T>> for FilteredAccessSet<T> {
 }
 
 impl<T: SparseSetIndex> FilteredAccess<T> {
+    /// Returns a `FilteredAccess` with no access and matching nothing.
+    /// This is the equivalent of a `FALSE` logic atom.
```
[the remark this answers is not in the document]
Comment (SkiFire13) [reply crates/bevy_ecs/src/query/access.rs:363]: That's a good idea!

**Review by Guvante — APPROVED** [review]
Does this change the behavior of `Or<()>`? The old code seems to be a no-op while the new code seems to remove all `filter_sets`. Mostly from a breaking changes standpoint.

**Comment by SkiFire13** [comment]
> Does this change the behavior of `Or<()>`? The old code seems to be a no-op while the new code seems to remove all `filter_sets`. Mostly from a breaking changes standpoint.

Yes, this is already mentioned in the migration guide.

I could avoid this breaking change by adding a bit of complexity in the code, but the current behaviour seems wrong to me so this is also an occasion to fix it. `Or` should match those archetypes that satisfy one of its inner filters, but if there's no filter then no filter is satisfied and no archetype should be matched, while the current behaviour is to match all of them. That said, I don't think someone is actually using `Or<()>` since it makes no sense, but if they are then chances are they expect the new behaviour, not the old one.
